### ISSUE TITLE
Spack is Great, but too slow

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#676]
Any ideas on what is causing the slowness, and how to speed it up?

**Comment by tgamblin** [comment]
Did [this commit](https://github.com/LLNL/spack/commit/b0b882cbb3d903a85413acb0439a02d6ac9e26fc) (from yesterday) help?

**Comment by citibeth** [comment]
Thanks, I will try that.

I haven't done any investigation myself.  But I've noticed Spack takes a
LONG time to chew through the specs I give it.  Even simple commands like
"spack spec" can take 10-15 seconds.

My overall idea starting this thread was to look for places where Spack
doesn't scale, and think through ways to fix them.  I think this is typical
that performance issues are uncovered when people start to use a project on
larger problems.

One potential scaling issue I've found: I noticed that Spack reads through
all the package.py files every time you start it up.

**Comment by tgamblin** [comment]
Yes, that's the most time consuming part.  See [this comment](https://github.com/LLNL/spack/pull/549#issuecomment-197069018):

> FYI: the big performance hit here comes from reading in all the packages to figure out what provides what.  You have to do that (currently) if you resolve a spec that has a virtual dep.  
> 
> For example, running `spack spec hypre` on the mainline with 329 packages:
> 
> ```bash
> $ time spack spec hypre &> /dev/null
> 
> real	0m3.834s
> user	0m3.532s
> sys	0m0.291s
> ```
> 
> And running the *exact* same command with a minimal repository that only has 11 packages:
> ```bash
> (atala):~$ time spack spec hypre &> /dev/null
> 
> real	0m0.582s
> user	0m0.444s
> sys	0m0.084s
> ```

> The obvious solution would be to index the package repo somehow, though you'd want to be able to detect any changes quickly and reindex based on what they are.  It would be annoying if you couldn't just hack `package.py` files and have the changes show up. 

The I referenced addresses this somewhat by reducing the time spent in directives, which now call `inspect.stack(0)` instead of `inspect.stack()`.  Directives only need access to the enclosing package scope, not the source code for each related frame, so we can pass `0` to skip the slow part.  I think there are still some bottlenecks in `import`, particularly on things like `NFS` mounts.

If we can reduce the time spent reading, the next bottleneck may be constraint solving.  I am not actually sure that takes a lot of time (or whether it ever will) because Spack specs are pretty small graphs.  But it's something to think about.  Luckily there are lots of solutions for fast constraint solvers, but it would be a significant rework of `spec.py`.

**Comment by citibeth** [comment]
> The obvious solution would be to index the package repo somehow,

I always thought the current virtual spec implementation might cause that
problem.  The kind of index needed would be a file for a virtual spec
indicating what implements it.  That could be implemented by hand, or
through some machine scanning process.

To be specific: mpi/package.py would be a virtual spec listing that it's
implemented by openmpi, mpich, etc.

**Comment by tgamblin** [comment]
We could do it that way, but I'd rather periodically generate something from scanning.  I really don't want to require vspecs to list their providers.  In fact, this won't work if you consider that Spack can have external package repositories.  a vspec may be provided by something in another package repo that Spack's builtin repo cannot know about.  This is something you want to maintain by hand.

I think there is a lot of room to optimize the reading process before we need to think too hard about indexing.  Right now there are extra metadata/stat operations that could be removed.  Running just `stat` on every file in the builtin repo only takes ~1.5sec even on NFS.

**Comment by eschnett** [comment]
There could be an "interpreter mode", where Spack starts once, and then accepts a series of (interactive) commands. This eliminates the start-up overhead if there are a series of commands. I often use Spack in this way, e.g. to get paths to several installed packages.

**Comment by citibeth** [comment]
>
> There could be an "interpreter mode", where Spack starts once, and then
> accepts a series of (interactive) commands. This eliminates the start-up
> overhead if there are a series of commands. I often use Spack in this way,
> e.g. to get paths to several installed packages.
>
Or better yet... Have a Spack command that does the Spack start-up, and
then dumps the resulting data structure to disk.  Subsequent Spack commands
could use that data structure for quick start-up.

Even better.... put that data structure in a .spack directory, and have a
quick procedure to determine whether it needs to be re-generated.
Basically, compare the size & timestamps of the relevant files in the Spack
directory, vs. what they were when the Spack init file was regenerated.

**Comment by tgamblin** [comment]
It wouldn't be hard to do that, but it would be nice if it also were easy to use on the command line.

**Comment by tgamblin** [comment]
> Even better.... put that data structure in a .spack directory, and have a
quick procedure to determine whether it needs to be re-generated.
Basically, compare the size & timestamps of the relevant files in the Spack
directory, vs. what they were when the Spack init file was regenerated.

There is AFAIK no particularly quick procedure to determine whether the index would need to be generated.  You will need to spider the `package.py` files and check their modification dates w.r.t. the index, so still `O(n)` FS operations.  Most filesystems do not change the modification dates of enclosing directories when files are modified.  Still, that scan would probably cost much less than the whole import.

How to cache the `package.py` files, and what to cache from them, is another question.  I think we should keep that minimal, *i.e.* maybe just start with a provider index.  I think there will be complications if we start trying to cache the entire `package.py` (and pickling is slow).

**Comment by citibeth** [comment]
>
> There is AFAIK no particularly quick procedure to determine whether the
> index would need to be generated. You will need to spider the package.py
> files and check their modification dates w.r.t. the index, so still O(n)
> FS operations. Most filesystems do not change the modification dates of
> enclosing directories when files are modified. Still, that scan would
> probably cost much less than the whole import.
>

Maybe this is a good start... I created a makefile (enclosed) with one
target, containing all 375 package.py and .patch files currently in the
Spack packages repo.  On my computer, `make` takes 1/8 second to determine
whether or not the target needs to be regenerated.

If `make` can do it, then so can Spack.  Maybe the way to go is
(internally) do an `ls -lR` of the directory, construct a makefile
on-the-fly, and call `make` on it.  If being fast about this involves lots
of clever OS-specific tricks, that way we don't have to reinvent them.

-- Elizabeth

mytarget: packages/py-pyyaml/package.py packages/swig/package.py packages/mpfr/package.py packages/ncurses/package.py packages/lwgrp/package.py packages/icu4c/package.py packages/py-netcdf/package.py packages/exuberant-ctags/package.py packages/py-pandas/package.py packages/zeromq/package.py packages/tau/package.py packages/cblas/package.py packages/harfbuzz/package.py packages/cleverleaf/package.py packages/lmod/package.py packages/lwm2/package.py packages/libffi/package.py packages/rose/package.py packages/rose/add_spack_compiler_recognition.patch packages/parallel-netcdf/package.py packages/stat/package.py packages/stat/configure_mpicxx.patch packages/munge/package.py packages/libedit/package.py packages/py-mx/package.py packages/py-virtualenv/package.py packages/opari2/package.py packages/triangle/package.py packages/espresso/package.py packages/pidx/package.py packages/xcb-proto/package.py packages/openspeedshop/package.py packages/libevent/package.py packages/scotch/package.py packages/libuuid/package.py packages/muster/package.py packages/cityhash/package.py packages/dyninst/package.py packages/py-yapf/package.py packages/autoconf/package.py packages/py-shiboken/package.py packages/automaded/package.py packages/judy/package.py packages/mpileaks/package.py packages/py-giss/package.py packages/pcre/package.py packages/emacs/package.py packages/activeharmony/package.py packages/gtkplus/package.py packages/spindle/package.py packages/libmng/package.py packages/protobuf/package.py packages/postgresql/package.py packages/py-pil/package.py packages/matio/package.py packages/py-pyqt/package.py packages/llvm/package.py packages/eigen/package.py packages/flex/package.py packages/qrupdate/package.py packages/cairo/package.py packages/gromacs/package.py packages/py-pillow/package.py packages/otf2/package.py packages/curl/package.py packages/gperf/package.py packages/py-pytables/package.py packages/gflags/package.py packages/libgd/package.py packages/py-numexpr/package.py packages/blitz/package.py packages/libpciaccess/package.py packages/adept-utils/package.py packages/arpack-ng/package.py packages/arpack-ng/pdlamch10.patch packages/automake/package.py packages/py-pyelftools/package.py packages/p4est/package.py packages/py-pycparser/package.py packages/ppl/package.py packages/vtk/package.py packages/readline/package.py packages/dri2proto/package.py packages/py-funcsigs/package.py packages/hpx5/package.py packages/exodusii/package.py packages/exodusii/exodus-cmake.patch packages/libdrm/package.py packages/otf/package.py packages/qhull/package.py packages/qhull/qhull-iterator.patch packages/dakota/package.py packages/oce/package.py packages/cbtf-krell/package.py packages/gnuplot/package.py packages/sundials/package.py packages/boost/package.py packages/boost/boost_11856.patch packages/py-lockfile/package.py packages/texinfo/package.py packages/metis/install_gklib_defs_rename.patch packages/metis/package.py packages/mvapich2/ad_lustre_rwcontig_open_source.patch packages/mvapich2/package.py packages/petsc/package.py packages/libtool/package.py packages/memaxes/package.py packages/cube/package.py packages/py-pylint/package.py packages/SAMRAI/package.py packages/SAMRAI/no-tool-build.patch packages/libNBC/package.py packages/cmocka/package.py packages/cryptopp/package.py packages/paraview/package.py packages/libsodium/package.py packages/ravel/package.py packages/netcdf-cxx4/package.py packages/libunwind/package.py packages/tetgen/package.py packages/py-six/package.py packages/task/package.py packages/fish/package.py packages/py-scientificpython/package.py packages/glpk/package.py packages/bzip2/package.py packages/libxshmfence/package.py packages/ncdu/package.py packages/py-matplotlib/package.py packages/patchelf/package.py packages/libcircle/package.py packages/wxpropgrid/package.py packages/silo/package.py packages/extrae/package.py packages/m4/package.py packages/m4/pgi.patch packages/libjson-c/package.py packages/openblas/package.py packages/jemalloc/package.py packages/tk/package.py packages/gcc/package.py packages/ompss/package.py packages/gdk-pixbuf/package.py packages/nasm/package.py packages/libdwarf/package.py packages/zlib/package.py packages/cfitsio/package.py packages/py-ipython/package.py packages/py-mako/package.py packages/qt/qt3krell.patch packages/qt/package.py packages/py-twisted/package.py packages/py-pychecker/package.py packages/py-sphinx/package.py packages/fftw/package.py packages/gasnet/package.py packages/elpa/package.py packages/tcl/package.py packages/gdb/package.py packages/asciidoc/package.py packages/jpeg/package.py packages/atop/package.py packages/ninja/package.py packages/Mitos/package.py packages/fltk/font.patch packages/fltk/package.py packages/vim/package.py packages/damselfly/package.py packages/cbtf-argonavis/package.py packages/graphviz/package.py packages/py-phonopy/package.py packages/xz/package.py packages/gnutls/package.py packages/parpack/package.py packages/py-biopython/package.py packages/mpe2/mpe2.patch packages/mpe2/package.py packages/cram/package.py packages/py-pexpect/package.py packages/icu/package.py packages/zsh/package.py packages/py-urwid/package.py packages/ndiff/package.py packages/parmetis/package.py packages/parmetis/enable_external_metis.patch packages/parmetis/pkg-parmetis-1c1a9fd0f408dc4d42c57f5c3ee6ace411eb222b.patch packages/parmetis/pkg-parmetis-82409d68aa1d6cbc70740d0f35024aae17f7d5cb.patch packages/wget/package.py packages/py-astropy/package.py packages/icebin/package.py packages/libjpeg-turbo/package.py packages/coreutils/package.py packages/yasm/package.py packages/python/package.py packages/py-numpy/package.py packages/netcdf-fortran/package.py packages/libsigsegv/package.py packages/fontconfig/package.py packages/julia/package.py packages/julia/gc.patch packages/cereal/package.py packages/cereal/Werror.patch packages/mesa/package.py packages/bear/package.py packages/bib2xhtml/package.py packages/octave/package.py packages/scorep/package.py packages/tmux/package.py packages/isl/package.py packages/libxml2/package.py packages/global/package.py packages/py-cython/package.py packages/szip/package.py packages/libxslt/package.py packages/py-pmw/package.py packages/libpng/package.py packages/gmsh/package.py packages/openssl/package.py packages/cgal/package.py packages/gmp/package.py packages/qthreads/package.py packages/ruby/package.py packages/snappy/package.py packages/valgrind/package.py packages/py-pyside/package.py packages/rsync/package.py packages/ImageMagick/package.py packages/py-pytz/package.py packages/flux/package.py packages/apex/package.py packages/py-sip/package.py packages/udunits2/package.py packages/bison/package.py packages/R/package.py packages/py-dateutil/package.py packages/py-mysqldb1/package.py packages/py-gnuplot/package.py packages/freetype/package.py packages/py-pyparsing/package.py packages/py-h5py/package.py packages/paraver/package.py packages/callpath/package.py packages/doxygen/package.py packages/netlib-lapack/package.py packages/pixman/package.py packages/libcerf/package.py packages/py-scikit-learn/package.py packages/git/package.py packages/netcdf/package.py packages/cppcheck/package.py packages/tmuxinator/package.py packages/pcre2/package.py packages/glib/package.py packages/cloog/package.py packages/openmpi/ad_lustre_rwcontig_open_source.patch packages/openmpi/package.py packages/openmpi/llnl-platforms.patch packages/openmpi/configure.patch packages/py-pygments/package.py packages/uncrustify/package.py packages/bowtie2/package.py packages/bowtie2/bowtie2-2.5.patch packages/netgauge/package.py packages/samtools/samtools1.2.patch packages/samtools/package.py packages/libxcb/package.py packages/hdf/package.py packages/gsl/package.py packages/mrnet/package.py packages/py-sympy/package.py packages/py-setuptools/package.py packages/ghostscript/package.py packages/libgpg-error/package.py packages/nettle/package.py packages/opencv/package.py packages/elfutils/package.py packages/util-linux/package.py packages/lcms/package.py packages/wx/package.py packages/trilinos/package.py packages/spot/package.py packages/pism/package.py packages/py-wheel/package.py packages/everytrace/package.py packages/glog/package.py packages/py-coverage/package.py packages/atlas/package.py packages/suite-sparse/package.py packages/py-tappy/package.py packages/pdt/package.py packages/lua/package.py packages/zfp/package.py packages/boxlib/package.py packages/scalasca/package.py packages/cbtf/package.py packages/glm/package.py packages/sparsehash/package.py packages/cbtf-lanl/package.py packages/py-periodictable/package.py packages/superlu-dist/package.py packages/tbb/package.py packages/hwloc/package.py packages/hypre/package.py packages/libelf/package.py packages/expat/package.py packages/cgm/package.py packages/cscope/package.py packages/lmdb/package.py packages/thrift/package.py packages/py-rpy2/package.py packages/gl2ps/package.py packages/gperftools/package.py packages/docbook-xml/package.py packages/py-blessings/package.py packages/arpack/package.py packages/mbedtls/package.py packages/xerces-c/package.py packages/mpc/package.py packages/pkg-config/package.py packages/issm/package.py packages/libgcrypt/package.py packages/py-genders/package.py packages/mpich/package.py packages/caliper/package.py packages/mumps/package.py packages/hdf5/package.py packages/py-proj/package.py packages/py-basemap/package.py packages/slepc/package.py packages/dbus/package.py packages/leveldb/package.py packages/py-cffi/package.py packages/atk/package.py packages/tar/package.py packages/proj/package.py packages/py-mpi4py/package.py packages/pmgr_collective/package.py packages/dtcmp/package.py packages/cmake/package.py packages/netlib-scalapack/package.py packages/taskd/package.py packages/libmonitor/libmonitorkrell-0002.patch packages/libmonitor/libmonitorkrell-0000.patch packages/libmonitor/package.py packages/libmonitor/libmonitorkrell-0001.patch packages/sqlite/package.py packages/jdk/package.py packages/ompt-openmp/package.py packages/binutils/package.py packages/binutils/cr16.patch packages/binutils/binutilskrell-2.24.patch packages/py-epydoc/package.py packages/scr/package.py packages/czmq/package.py packages/py-pbr/package.py packages/libtiff/package.py packages/py-nose/package.py packages/py-scipy/package.py packages/ibmisc/package.py packages/graphlib/package.py packages/llvm-lld/package.py packages/launchmon/package.py packages/geos/package.py packages/libarchive/package.py packages/mpibash/package.py packages/mpibash/mpibash-4.3.patch packages/googletest/package.py packages/the_silver_searcher/package.py packages/gettext/package.py packages/py-pypar/package.py packages/muparser/package.py packages/pango/package.py packages/py-mock/package.py packages/papi/package.py packages/py-python-daemon/package.py 
	date > mytarget

**Comment by eschnett** [comment]
There's `spack reindex` because Spack doesn't look at all installed packages all the time. There could be a similar `spack relist` to spider the packages. The faster intermediate format could be as simple as a yaml file mapping package names and dependencies to their files.

**Comment by tgamblin** [comment]
`reindex` is similar, but it *should* never have to run, if there are no bugs.  It's there in case your database gets corrupted.  Here, reindexing needs to happen more often.  But yes, adding a command makes sense.

I personally like the idea of adding options to reindex rather than adding a separate `relist`.  Or maybe `spack repo reindex <repo>` where `<repo>` defaults to the builtin one, or all registered repos where the user has write permission.

+1 for yaml

**Comment by citibeth** [comment]
I tried `spack --profile`  For example:
```
$ spack --profile uninstall -a glint2
```
took 33 seconds.  Here are the top functions. ordered by internal time:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
7690385/55579    3.225    0.000   16.421    0.000 {hash}
820953/6174    2.383    0.000   16.523    0.003 spec.py:1654(_cmp_key)
3614581/53774    2.323    0.000   16.391    0.000 lang.py:248(<lambda>)
1658880/1591584    1.942    0.000    4.901    0.000 {sorted}
   113784    1.762    0.000    2.106    0.000 emitter.py:629(analyze_scalar)
2143389/2109162    1.564    0.000    3.021    0.000 lang.py:241(<lambda>)
  5455941    1.172    0.000    1.172    0.000 spec.py:304(_cmp_key)
   112117    1.118    0.000    1.591    0.000 emitter.py:1082(write_plain)
  1629621    1.005    0.000    2.407    0.000 version.py:717(__hash__)
146408/6410    0.977    0.000   11.235    0.002 serializer.py:78(serialize_node)
3542289/3542288    0.891    0.000    0.891    0.000 {isinstance}
1635732/59938    0.888    0.000   16.478    0.000 spec.py:1662(<genexpr>)
   814859    0.741    0.000    4.530    0.000 lang.py:257(_cmp_key)
   204672    0.586    0.000    9.712    0.000 emitter.py:111(emit)
  1629605    0.570    0.000    0.755    0.000 version.py:249(__hash__)
    69999    0.538    0.000    0.699    0.000 emitter.py:582(prepare_tag)
   146408    0.474    0.000    4.465    0.000 emitter.py:233(expect_node)
   396461    0.411    0.000    0.411    0.000 {method 'encode' of 'unicode' objects}
```
Most surprising to me is, why are some functions being called 10s of millions of times?

**Comment by adamjstewart** [comment]
I'm also seeing slow times for complicated Python packages. [py-sncosmo](#1044) took 21 minutes to uninstall:
```
[09:50:45] ajstewart@blogin4:/soft/spack (features/sncosmo *%)
$ spack -v -p uninstall py-sncosmo
==> The following packages will be uninstalled : 

-- linux-x86_64 / gcc@6.1.0 -------------------------------------
426qlx2 py-sncosmo@1.2.0%gcc


==> Do you want to proceed ? [y/n]
y
==> Successfully uninstalled py-sncosmo@1.2.0%gcc@6.1.0 arch=linux-x86_64-426qlx2
Traceback (most recent call last):
  File "/soft/spack/bin/spack", line 171, in <module>
    cProfile.run('main()', sort='tottime')
  File "/usr/lib64/python2.6/cProfile.py", line 36, in run
    result = prof.print_stats(sort)
  File "/usr/lib64/python2.6/cProfile.py", line 81, in print_stats
    pstats.Stats(self).strip_dirs().sort_stats(sort).print_stats()
  File "/usr/lib64/python2.6/pstats.py", line 232, in sort_stats
    sort_tuple = sort_tuple + sort_arg_defs[word][0]
KeyError: 'tottime'
[10:11:16] ajstewart@blogin4:/soft/spack (features/sncosmo *%)
$ 
```
It also looks like there's a bug in the profiler.

**Comment by glennpj** [comment]
I am experiencing poor, and variable, performance with spack. For instance a `spack find` can take anywhere from 10 to 20 minutes. A `spack module refresh` to add a suffix string took ~45 minutes. I had an uninstall running for about 3 hours before I finally killed the process. I am working in a VM so that does not help but it seems running any spack commands that query installed packages takes a really long time. I have 235 packages installed and I noticed things slowing down at around 100. I tried PR #1015 but that did not seem to help. There were conflicts with dependency types so it is possible that I did not get the merge quite right.
